### PR TITLE
DEVELOPMENT-GUIDE: add Development Guide 

### PR DESCRIPTION
**Pull request by alltilla**
Intends to do two things:
* Give new contributors a high level view of the source code and architecture.
* Document our agreed but not written coding conventions.

Signed-off-by: Attila Szakacs <attila.szakacs@axoflow.com>

### CODE REVIEW

File: DEVELOPMENT-GUIDE.md [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+The code is organized into:
+- Core library (`lib` directory).
+- Plugin drivers (`modules` directory).
+
+The configuration language is defined in `lib/cfg-grammar.y` and module related `...-grammar.ym` files.
+
+Supported build systems are Autotools and CMake.
+
+Testing:
+- Unit tests use the [Criterion][ar:criterion] framework, and can be found alongside its respective source code (`.../tests` directory).
```
Comment (OverOrion) [inline DEVELOPMENT-GUIDE.md:20]:
```suggestion
- Unit tests use the [Criterion][ar:criterion] framework and can be found alongside its respective source code (`.../tests` directory).
```

File: DEVELOPMENT-GUIDE.md [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+### Git history
+
+- Each commit is atomic and detailed in the commit message.
+- Commits represent the logical buildup of a feature, organized via interactive rebase to improve readability and maintainability.
+- For more info see [Commits](CONTRIBUTING.md#commits).
+
+### Comments
+
+- Comments are rarely used, as clean coding and atomic commits are the main way to document the code.
```
Comment (OverOrion) [inline DEVELOPMENT-GUIDE.md:69]:
```suggestion
- Comments are rarely used, as clean coding and atomic commits are the main ways to document the code.
```

File: DEVELOPMENT-GUIDE.md [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+- Functionality is layered in abstractions to enhance readability and maintainability.
+- Function and variable names are descriptive to accurately communicate each layer's purpose.
+
+### C code format
+
+AxoSyslog follows GNU standards with additional conventions:
+- Classes are in PascalCase, functions and variables in snake_case.
+- Class namespaces are prefixed in function names.
+- Private methods are `static` and start with an underscore (`_`).
+- Function signatures are split to two lines (return type on the first, rest on the second) to help `git grep ^my_func_impl` searches.
```
Comment (OverOrion) [inline DEVELOPMENT-GUIDE.md:58]:
```suggestion
- Function signatures are split into two lines (return type on the first, rest on the second) to help `git grep ^my_func_impl` searches.
```

**Review by OverOrion — COMMENTED** [review]
LGTM, some minor typos only, thank you!

File: DEVELOPMENT-GUIDE.md (hunk 1 above)
Comment (alltilla) [reply DEVELOPMENT-GUIDE.md]: Thanks, fixed.

File: DEVELOPMENT-GUIDE.md (hunk 2 above)
Comment (alltilla) [reply DEVELOPMENT-GUIDE.md]: Thanks, fixed.

File: DEVELOPMENT-GUIDE.md (hunk 3 above)
Comment (alltilla) [reply DEVELOPMENT-GUIDE.md]: Thanks, fixed.

**Review by MrAnno — APPROVED** [review]

**Review by OverOrion — APPROVED** [review]

File: DEVELOPMENT-GUIDE.md [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+- Core library (`lib` directory).
+- Plugin drivers (`modules` directory).
+
+The configuration language is defined in `lib/cfg-grammar.y` and module related `...-grammar.ym` files.
+
+Supported build systems are Autotools and CMake.
+
+Testing:
+- Unit tests use the [Criterion][ar:criterion] framework and can be found alongside its respective source code (`.../tests` directory).
+- End-to-end tests are written in Python with `pytest` (`tests/light` directory).
```
Comment (mitzkia) [inline DEVELOPMENT-GUIDE.md:21]: Maybe here we can use GitHub URL reference for pytest as well - https://github.com/pytest-dev/pytest

Comment (alltilla) [reply DEVELOPMENT-GUIDE.md]: Thanks, fixed!

Comment (mitzkia) [reply DEVELOPMENT-GUIDE.md]: Thanks for the fix

**Review by mitzkia — APPROVED** [review]

**Pull request merged**